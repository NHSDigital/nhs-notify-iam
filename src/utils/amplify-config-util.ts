import { Amplify, ResourcesConfig } from 'aws-amplify';
import {
  CIS2_PROVIDER_NAME,
  COGNITO_DOMAIN,
  REDIRECT_DOMAIN,
  USER_POOL_CLIENT_ID,
  USER_POOL_ID,
} from './constants';

// eslint-disable-next-line @typescript-eslint/no-require-imports, unicorn/prefer-module, import/no-unresolved
const amplifyOutputs = require('@/amplify_outputs.json');

console.log(amplifyOutputs);

export function configureAmplify() {
  const authConfig: ResourcesConfig = {
    Auth: {
      Cognito: {
        userPoolId: USER_POOL_ID,
        userPoolClientId: USER_POOL_CLIENT_ID,
        loginWith: {
          oauth: {
            domain: COGNITO_DOMAIN,
            scopes: [
              'aws.cognito.signin.user.admin',
              'email',
              'openid',
              'profile',
            ],
            redirectSignIn: [REDIRECT_DOMAIN],
            redirectSignOut: [],
            responseType: 'code',
            providers: [{ custom: CIS2_PROVIDER_NAME }],
          },
        },
      },
    },
  };
  Amplify.configure({ ...amplifyOutputs }, { ssr: true });
}
