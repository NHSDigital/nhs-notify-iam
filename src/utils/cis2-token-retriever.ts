'use server';

import axios from 'axios';
import { Amplify } from 'aws-amplify';

export type TokenResponse = {
  id_token: string;
  access_token: string;
  refresh_token: string;
  expires_in: number;
  token_type: 'Bearer';
};

// eslint-disable-next-line @typescript-eslint/no-require-imports, unicorn/prefer-module, import/no-unresolved
Amplify.configure(require('@/amplify_outputs.json'), { ssr: true });

export async function getAuthTokens(
  code: string,
  codeVerifier: string
): Promise<TokenResponse | undefined> {
  if (!code || !codeVerifier) {
    return;
  }

  const amplifyConfig = Amplify.getConfig();

  const cognitoConfig = amplifyConfig.Auth?.Cognito;
  const userPoolClientId = cognitoConfig?.userPoolClientId || '';
  const redirectDomain = (cognitoConfig?.loginWith?.oauth?.redirectSignIn ||
    [])[0];
  const cognitoDomain = cognitoConfig?.loginWith?.oauth?.domain;

  const params = new URLSearchParams();
  params.append('grant_type', 'authorization_code');
  params.append('code', code);
  params.append('client_id', userPoolClientId);
  params.append('redirect_uri', redirectDomain);
  params.append('code_verifier', codeVerifier);

  const response = await axios.post<TokenResponse>(
    `https://${cognitoDomain}/oauth2/token`,
    params,
    {
      auth: {
        username: userPoolClientId,
        password: process.env.USER_POOL_CLIENT_SECRET || '',
      },
    }
  );
  return response.data;
}
