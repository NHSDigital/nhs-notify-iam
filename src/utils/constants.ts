// client and server
export const BASE_PATH = process.env.NEXT_PUBLIC_BASE_PATH ?? '/auth';
export const USER_POOL_ID = process.env.NEXT_PUBLIC_USER_POOL_ID || '';
export const USER_POOL_CLIENT_ID =
  process.env.NEXT_PUBLIC_USER_POOL_CLIENT_ID || '';
export const COGNITO_DOMAIN = process.env.NEXT_PUBLIC_COGNITO_DOMAIN || '';
export const REDIRECT_DOMAIN = process.env.NEXT_PUBLIC_REDIRECT_DOMAIN || '';
export const CIS2_PROVIDER_NAME =
  process.env.NEXT_PUBLIC_CIS2_PROVIDER_NAME || '';

