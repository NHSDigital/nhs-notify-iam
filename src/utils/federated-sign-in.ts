import { signInWithRedirect } from '@aws-amplify/auth';
import { AmplifyOutputs } from 'aws-amplify/adapter-core';

// eslint-disable-next-line @typescript-eslint/no-require-imports, unicorn/prefer-module, import/no-unresolved
const amplifyOutputs: AmplifyOutputs = require('@/amplify_outputs.json');

export type State = {
  redirectPath: string;
};

export function federatedSignIn(redirectPath: string) {
  const providers = amplifyOutputs.auth?.oauth?.identity_providers || [];
  if (providers.length !== 1) {
    throw new Error('Invalid OAUTH provider configuration');
  }
  const provider = providers[0];
  if (!provider) {
    throw new Error('Missing OAUTH custom provider configuration');
  }
  return signInWithRedirect({
    provider: {
      custom: provider,
    },
    customState: JSON.stringify({ redirectPath }),
  });
}
