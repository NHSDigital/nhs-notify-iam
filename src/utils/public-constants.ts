// client and server
export const getConstants: () => Record<string, string> = () => ({
  BASE_PATH: process.env.NEXT_PUBLIC_BASE_PATH ?? '/auth',
  USER_POOL_ID: process.env.NEXT_PUBLIC_USER_POOL_ID || '',
  USER_POOL_CLIENT_ID: process.env.NEXT_PUBLIC_USER_POOL_CLIENT_ID || '',
  COGNITO_DOMAIN: process.env.NEXT_PUBLIC_COGNITO_DOMAIN || '',
  REDIRECT_DOMAIN: process.env.NEXT_PUBLIC_REDIRECT_DOMAIN || '',
  CIS2_PROVIDER_NAME: process.env.NEXT_PUBLIC_CIS2_PROVIDER_NAME || '',
});

export const getServerConstants: () => Record<string, string> = () => ({
  ...getConstants(),
  ...{ USER_POOL_CLIENT_SECRET: process.env.USER_POOL_CLIENT_SECRET || '' },
});
