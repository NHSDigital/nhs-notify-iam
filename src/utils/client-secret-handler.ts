'use server';

import crypto from 'crypto';
import { USER_POOL_CLIENT_ID } from '@/src/utils/constants';

const userPoolClientSecret = process.env.USER_POOL_CLIENT_SECRET || '';

export async function generateClientSecretHash(
  username: string
): Promise<string> {
  const hasher = crypto.createHmac('sha256', userPoolClientSecret);
  hasher.update(username);
  hasher.update(USER_POOL_CLIENT_ID);
  return hasher.digest('base64');
}
