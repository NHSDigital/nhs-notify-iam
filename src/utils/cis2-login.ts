import { signInWithRedirect } from '@aws-amplify/auth';
import { getConstants } from '@/src/utils/public-constants';
import { ResourcesConfig } from 'aws-amplify';
import { AmplifyOutputs } from 'aws-amplify/adapter-core';

export type State = {
  redirectPath: string;
  requestRef: string;
};

// export function stateParser(state?: string | null): State | undefined {
//     if (!state) {
//         return undefined;
//     }

//         const parts = state.split('-');
//         console.log(parts);
//         if (parts.length > 1) {
//           const stateData = JSON.parse(Buffer.from(parts[1], 'hex').toString('utf-8'));
//             console.log(`stateData ${JSON.stringify(stateData)}`);
//             redirectPath = stateData['redirectPath'];
//         }
//       }
// }

export function cis2Login(redirectPath: string, requestRef: string) {
  const provider = getConstants().CIS2_PROVIDER_NAME;
  return signInWithRedirect({
    provider: {
      custom: provider,
    },
    customState: JSON.stringify({ redirectPath, requestRef }),
  });
}

export function getCis2AmplifyConfiguration(): AmplifyOutputs {
  const {
    USER_POOL_ID,
    USER_POOL_CLIENT_ID,
    COGNITO_DOMAIN,
    REDIRECT_DOMAIN,
    CIS2_PROVIDER_NAME,
  } = getConstants();

  return {
    version: '1.3', // version is important for the framework to recognise that this type of config is AmplifyOutputs
    auth: {
      aws_region: 'eu-west-2',
      user_pool_id: USER_POOL_ID,
      user_pool_client_id: USER_POOL_CLIENT_ID,
      oauth: {
        identity_providers: [CIS2_PROVIDER_NAME],
        domain: COGNITO_DOMAIN,
        scopes: ['aws.cognito.signin.user.admin', 'email', 'openid', 'profile'],
        redirect_sign_in_uri: [REDIRECT_DOMAIN],
        redirect_sign_out_uri: [],
        response_type: 'code',
      },
    },
  };
}
