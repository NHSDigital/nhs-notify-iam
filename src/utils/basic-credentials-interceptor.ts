// Hard to avoid due to FetchInterceptor using any
/* eslint @typescript-eslint/no-explicit-any: 0 */
import { FetchInterceptor } from 'fetch-intercept';
import { generateClientSecretHash } from './client-secret-handler';

const TARGET_INITIATE = 'AWSCognitoIdentityProviderService.InitiateAuth';
const TARGET_RESPONSE_TO_CHALLENGE =
  'AWSCognitoIdentityProviderService.RespondToAuthChallenge';

const fetchInterceptTargets = new Set([
  TARGET_INITIATE,
  TARGET_RESPONSE_TO_CHALLENGE,
]);

function extractUsername(target: string, body: any): string {
  if (target === TARGET_INITIATE) {
    return body.AuthParameters.USERNAME;
  }
  if (target === TARGET_RESPONSE_TO_CHALLENGE) {
    return body.ChallengeResponses.USERNAME;
  }
  return '';
}

function injectSecretHash(target: string, body: any, secretHash: string): any {
  if (target === TARGET_INITIATE) {
    body.AuthParameters.SECRET_HASH = secretHash;
  } else if (target === TARGET_RESPONSE_TO_CHALLENGE) {
    body.ChallengeResponses.SECRET_HASH = secretHash;
  }
  return body;
}

export const basicCredentialsInterceptor: FetchInterceptor = {
  request: (url: string, config: any) => {
    if (!config?.headers) {
      return [url, config];
    }

    const target = config.headers['x-amz-target'] || '';
    if (!fetchInterceptTargets.has(target)) {
      return [url, config];
    }

    const body = JSON.parse(config.body);
    const username = extractUsername(target, body);

    return generateClientSecretHash(username).then((secretHash) => {
      const updatedBody = injectSecretHash(target, body, secretHash);
      config.body = JSON.stringify(updatedBody);
      return [url, config];
    });
  },
};
