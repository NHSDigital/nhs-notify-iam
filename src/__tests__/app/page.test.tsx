import { act, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Hub } from 'aws-amplify/utils';
import { redirect } from 'next/navigation';
import { federatedSignIn } from '@/src/utils/federated-sign-in';
import SignInPage from '@/src/app/page';

jest.mock('aws-amplify/utils', () => ({
  Hub: {
    listen: jest.fn(),
  },
}));

const mockGetSearchParams = jest.fn();

jest.mock('next/navigation', () => ({
  redirect: jest.fn(),
  RedirectType: { replace: 'replace' },
  useSearchParams: () => ({
    get: mockGetSearchParams,
  }),
}));

jest.mock('@aws-amplify/ui-react', () => ({
  Authenticator: () => <div>Placeholder Sign-in form</div>,
}));

jest.mock('@/src/components/CIS2SignInButton/CIS2SignInButton', () => ({
  CIS2SignInButton: ({ onClick }: { onClick: () => void }) => (
    <button type='button' data-testid='mock-cis2-button' onClick={onClick}>
      Mock CIS2 Sign In Button
    </button>
  ),
}));

jest.mock('@/src/utils/federated-sign-in');

const mockedHub = jest.mocked(Hub);
const mockedRedirect = jest.mocked(redirect);
const mockFederatedSignIn = jest.mocked(federatedSignIn);

function getEventListener() {
  const lastHubListenCall = mockedHub.listen.mock.lastCall;
  const eventType = lastHubListenCall?.[0];
  const eventListener = lastHubListenCall?.[1];

  expect(eventType).toBe('auth');

  expect(eventListener).toBeTruthy();
  return eventListener!;
}

describe('SignInPage', () => {
  const originalCognitoIdpSetting = process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP;

  it('renders', async () => {
    process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP = 'false';

    const container = render(<SignInPage />);

    expect(container.asFragment()).toMatchSnapshot();

    process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP = originalCognitoIdpSetting;
  });

  it('renders with cognito login form if env var switch is enabled', () => {
    process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP = 'true';

    const container = render(<SignInPage />);

    expect(container.asFragment()).toMatchSnapshot();

    process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP = originalCognitoIdpSetting;
  });

  it('does federated sign in when clicking cis2 button', async () => {
    const user = userEvent.setup();

    const container = render(<SignInPage />);

    const button = container.getByTestId('mock-cis2-button');

    await user.click(button);

    expect(mockFederatedSignIn).toHaveBeenCalledWith('/home');
  });

  it('sets redirect based on search params when clicking cis2 button', async () => {
    const user = userEvent.setup();

    mockGetSearchParams.mockReturnValueOnce('/example-redirect');

    const container = render(<SignInPage />);

    const button = container.getByTestId('mock-cis2-button');

    await user.click(button);

    expect(mockFederatedSignIn).toHaveBeenCalledWith('/example-redirect');
  });

  it('listens for oauth state events', async () => {
    const page = <SignInPage />;
    const container = render(page);

    const eventListener = getEventListener();

    act(() => {
      eventListener({
        payload: {
          event: 'customOAuthState',
          data: '{"redirectPath":"/testing"}',
        },
        channel: '',
      });
      container.rerender(page);
    });

    await waitFor(() =>
      expect(mockedRedirect).toHaveBeenCalledWith(
        '?redirect=%2Ftesting',
        'replace'
      )
    );
  });

  it('ignores other state events', async () => {
    const page = <SignInPage />;
    const container = render(page);

    const eventListener = getEventListener();

    act(() => {
      eventListener({
        payload: {
          event: 'ignoreThis',
          data: '{"redirectPath":"/testing"}',
        },
        channel: '',
      });
      container.rerender(page);
    });

    await waitFor(() => expect(mockedRedirect).not.toHaveBeenCalled());
  });

  it('defaults redirect path', async () => {
    const page = <SignInPage />;
    const container = render(page);

    const eventListener = getEventListener();

    act(() => {
      eventListener({
        payload: {
          event: 'customOAuthState',
          data: '{"redirectPath":""}',
        },
        channel: '',
      });
      container.rerender(page);
    });

    await waitFor(() =>
      expect(mockedRedirect).toHaveBeenCalledWith(
        '?redirect=%2Fhome',
        'replace'
      )
    );
  });
});
