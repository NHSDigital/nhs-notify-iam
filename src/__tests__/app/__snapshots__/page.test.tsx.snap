// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInPage renders 1`] = `
<DocumentFragment>
  <div>
    Placeholder Sign-in form
  </div>
</DocumentFragment>
`;
